### ISSUE TITLE
❓ Question about label format: empty .txt files and multiple instances per class

### ISSUE THREAD

**Issue by dvpalomino** [JDSobek/MedYOLO#38]
Hi, I have a question regarding the label format:

Is it acceptable to include empty label files?
For example, label001.txt exists but contains no annotations (i.e., it's an empty .txt file). Will this be correctly interpreted as "no object present" for that image?

Can a label file contain multiple instances of the same class?
For instance:

label001.txt
1 0.610294 0.244141 0.725586 0.044118 0.035156 0.037109 1.0000
1 0.294118 0.328125 0.719727 0.029412 0.023438 0.025391 1.0000
Both lines refer to objects of the same class (class 1). Is this format valid?

Thanks in advance!

**Comment by JDSobek** [comment]
Empty labels should be fine. IIRC they aren't necessary, if a label file doesn't exist for an image the framework should interpret that as "no objects in this image."

It should also be fine with multiple instances of the same class, but I never had the right dataset to test it. The YOLOv5 code it's based off of does detect multiple instances of the same class and nothing I did should have changed that... though this model can be finicky about detecting small objects which could confound any tests.

> Both lines refer to objects of the same class (class 1). Is this format valid?

Almost, the last 1.000000 shouldn't be there. The model will output a "confidence level" at the end of the line during inference, but the input labels shouldn't have it. Though looking through the code I think the model might just ignore anything after the 7th place on each line.

**Comment by dvpalomino** [comment]
Thanks a lot for the clarification! I’ll make sure to remove the confidence value from the ground truth labels.

Really appreciate your help!

**Comment by dvpalomino** [comment]
Hello again,

I’d like to know if training can be resumed after being interrupted — for example, due to a system shutdown or manual stop. 

1. Is there a way to resume training from the last checkpoint? If so, what’s the proper way to do it?

2. Where does the model store checkpoints or intermediate training information?


I'd like to make sure this data is being saved properly so I don’t lose progress.

Thanks in advance!

**Comment by JDSobek** [comment]
1.  [This setting should let you resume a training run](https://github.com/JDSobek/MedYOLO/blob/main/train.py#L370). However, I've never tested it and I'm not 100% sure how it works. [It picks this run, which I think is just time-indexed](https://github.com/JDSobek/MedYOLO/blob/main/utils/general.py#L34) or it looks like you can pass it a path to the weights you want to start with, but I'm not sure how it knows what epoch training stopped at. Otherwise, you might be able to [load your weights](https://github.com/JDSobek/MedYOLO/blob/main/train.py#L363) into a training run and [change the hyperparameter configuration](https://github.com/JDSobek/MedYOLO/blob/main/train.py#L366) to [a more appropriate set of hyperparameters for fine-tuning a partially trained model](https://github.com/JDSobek/MedYOLO/blob/main/train.py#L366) and train for however many epochs.
2. [This setting determines where weights are saved](https://github.com/JDSobek/MedYOLO/blob/main/train.py#L380). [This one determines the name of the folder for any given training run](https://github.com/JDSobek/MedYOLO/blob/main/train.py#L381).

**Comment by dvpalomino** [comment]
Thank you very much for the help! It's just what I was looking for.

**Comment by dvpalomino** [comment]
UPDATE: Hi! I'm trying to resume training from a previous checkpoint (last.pt). I checked the file manually using this snippet:

```
###############################################################
import torch

checkpoint_path = "D:/MedYOLO/runs/train/exp16/weights/last.pt"
checkpoint = torch.load(checkpoint_path, map_location='cpu')
last_epoch = checkpoint.get('epoch', None)

if last_epoch is not None:
    print(f"Last epoch saved: {last_epoch}")
else:
    print("No epochs found in the checkpoint")
###############################################################
```


So it seems that the --resume flag is automatically unpickling the last.pt checkpoint and reading the last epoch somehow like in the snippet of code. However, when i run the command:

```
###############################################################
python train.py --data LUNA_dataset.yaml --adam --norm CT --epochs 1000 --patience 200 --device 0 --batch-size 3 --resume 'D:\MedYOLO\runs\train\exp16\weights\last.pt' 
###############################################################
```

The train seems to start again from epoch 0 out of 999. Is this expected behavior? Shouldn’t --resume automatically continue from the last saved epoch if the experiment folder contains last.pt?

I see that in the train.py script the start epoch is hardcoded in line 155:
```
###############################################################
start_epoch, best_fitness = 0, 0.0
###############################################################
```

and the used in line 210:

```
###############################################################
scheduler.last_epoch = start_epoch - 1  # do not move
###############################################################
```

Let me know if I’m missing something or if there’s a better way to resume training. Thanks in advance!

**Comment by dvpalomino** [comment]
UPDATE: adding this snippet of code to the train.py script in line 157:

    ###############################################################
    #Path to checkpoint
    ckpt = opt.resume if isinstance(opt.resume, str) else get_latest_run()  # specified or most recent path
    print(f"checkpoint path: {ckpt}")

    #Load the checkpoint
    checkpoint = torch.load(ckpt, map_location='cpu')

    #Read last epoch and fitness
    last_epoch = checkpoint.get('epoch', None)
    last_fitness = checkpoint.get('best_fitness', None)

    start_epoch = int(last_epoch)
    best_fitness = float(last_fitness)

    if last_epoch is not None:
        print(f"Last saved epoch: {last_epoch}")
        print(f"Best fitness saved: {last_fitness}")
    else:
        print("No key 'epoch' or 'best_fitness' found in checkpoint")
    ###############################################################

and setting the appropiate epoch at the training loop inside train.py as follows:

    ###############################################################
    for epoch in range(epochs-start_epoch):  # epoch ------------------------------------------------------------------
        epoch = epoch + start_epoch
        .....
    ###############################################################

This worked for me  to resume an interrupted training from the last.pt checkpoint. Let me know if there is something I am not taking into account. Thank you!

**Comment by JDSobek** [comment]
Based on the old and current YOLOv5 commits, I think it's either intended to do that or it might be an artifact of me removing weights and biases functionality during the testing phase which it seems like they also used for resuming training runs. I don't see an issue with the changes you made and it makes more sense than what it's actually doing.

[MedYOLO origin YOLOv5 commit](https://github.com/ultralytics/yolov5/blob/ed887b5976d94dc61fa3f7e8e07170623dc7d6ee/train.py#L495)
[Current YOLOv5 commit](https://github.com/ultralytics/yolov5/blob/0c99ce80041eee7f2df500bcbdded754469a7124/train.py#L641)

Unfortunately I no longer have access to infrastructure for testing changes. I'll make a separate issue for your fix so we can keep it recorded for people who need it in the future.
